tests/cupy_tests/testing_tests/test_helper.py

```diff
@@ -71,6 +71,54 @@ def test_same_error(self):
                                        numpy_error, numpy_tb,
                                        accept_error=Exception)
 
+    def test_cupy_derived_error(self):
+        cupy_error = ValueError()
+        cupy_tb = 'xxxx'
+        numpy_error = Exception()
+        numpy_tb = 'yyyy'
+        # Nothing happens, CuPy errors may derive from NumPy errors
+        helper._check_cupy_numpy_error(self, cupy_error, cupy_tb,
```

Comment:
How about testing `helper.numpy_cupy_raises`? (Maybe you can use `@pytest.mark.xfail(raises=)`)?

Reply:
In the sense that we should use the pytest decorator in favor of `self.assert*`?

Reply:
My point is that we should test `helper.numpy_cupy_raises` instead of `helper._check_cupy_numpy_error` as long as possible, because the latter is internal implementation.
If we're going to test the former, I think we can use `xfail` like (not tested though):

```
@pytest.mark.xfail(raises=AssertionError)
@testing.numpy_cupy_raises(...)
def test_should_fail(xp, ...):
    raise incompatible errors
```

Reply:
Using `xfail` for testing expected exceptions is not recommended.

`... whereas using @pytest.mark.xfail with a check function is probably better for something like documenting unfixed bugs (where the test describes what “should” happen) or bugs in dependencies.`
https://docs.pytest.org/en/latest/assert.html

How about this approach. It address your point, does not directly depend on pytest, and can probably be applied to all fixtures of interest.

```
def test_cupy_derived_error(self):
    @testing.helper.numpy_cupy_raises(accept_error=Exception)
    def dummy_test_cupy_derived_error(self, xp):
        if xp is cupy:
            raise ValueError()
        elif xp is numpy:
            raise Exception()
    dummy_test_cupy_derived_error(self)
```

Reply:
>Using xfail for testing expected exceptions is not recommended.

Oh, thank you for pointing out!
That might work, but feels like a little bit unsafe, because `dummy_test_cupy_derived_error` in this code is a function instead of a bound method.

How about `pytest.raises`? I think it's ok to depend on pytest. At least alternative to this feature is easy to implement.

Reply:
Updated fixtures to define decorated functions, after offline discussions that decorated methods will become functions (unbounded) anyway.